crypto/compactcert/weights.go

```diff
@@ -0,0 +1,183 @@
+// This function checks that the following inequality is satisfied
+//
+// numReveals * (3 * 2^b * (signedWeight^2 - 2^2d) + d * (T-1) * Y) >= ((strengthTarget) * T + numReveals * P) * Y
+//
+// where signedWeight/(2^d) >=1 for some integer d>=0, p = P/(2^b) >= ln(provenWeight), t = T/(2^b) >= ln(2) >= (T-1)/(2^b)
+// for some integers P,T >= 0 and b=16.
+//
+// T and b are defined in the code as the constants ln2IntApproximation and precisionBits respectively.
+// P is set to lnProvenWeight argument
+// more details can be found on the Algorand's spec
+func verifyWeights(signedWeight uint64, lnProvenWeight uint64, numOfReveals uint64, strengthTarget uint64) error {
+	if numOfReveals > MaxReveals {
+		return ErrTooManyReveals
+	}
+
+	if signedWeight == 0 {
+		return ErrZeroSignedWeight
+	}
+
+	// in order to make the code more readable and reusable we will define the following expressions:
+	// y = signedWeight^2 + 2^(d + 2) * signedWeight + 2^2d
+	// x = 3 * 2^b * (signedWeight^2 - 2^2d)
+	// w = d * (T - 1)
+	//
+	//  numReveals * (3 * 2^b * (signedWeight^2 - 2^2d) + d * (T-1) * Y) >= ((strengthTarget) * T + numReveals * P) * Y
+	//        /\
+	//        ||
+	//        \/
+	// numReveals * (x + w * y >= ((strengthTarget) * T + numReveals * P) * y
+	y, x, w := getSubExpressions(signedWeight)
+	lhs := &big.Int{}
+	lhs.Set(w).
+		Mul(lhs, y).
+		Add(x, lhs).
+		Mul(bigInt(numOfReveals), lhs)
+
+	revealsTimesP := &big.Int{}
+	revealsTimesP.Set(bigInt(numOfReveals)).Mul(revealsTimesP, bigInt(lnProvenWeight))
+
+	rhs := &big.Int{}
+	rhs.Set(bigInt(strengthTarget))
+	rhs.Mul(rhs, bigInt(ln2IntApproximation)).
+		Add(rhs, revealsTimesP).
+		Mul(rhs, y)
+
+	if lhs.Cmp(rhs) < 0 {
+		return ErrInsufficientSingedWeight
+	}
+
+	return nil
+}
+
+// numReveals computes the number of reveals necessary to achieve the desired
+// security parameters. we use value which satisfies the following equation:
+//
+// numReveals > = ((strengthTarget) * T * Y / (3 * 2^b * (signedWeight^2 - 2^2d) + (d * (T - 1) - P) * Y))
+// where signedWeight/(2^d) >=1 for some integer d>=0, p = P/(2^b) >= ln(provenWeight), t = T/(2^b) >= ln(2) >= (T-1)/(2^b)
+// for some integers P,T >= 0 and b=16.
+//
+// T and b are defined in the code as the constants ln2IntApproximation and precisionBits respectively.
+// P is set to lnProvenWeight argument
+// more details can be found on the Algorand's spec
+func numReveals(signedWeight uint64, lnProvenWeight uint64, strengthTarget uint64) (uint64, error) {
+	// in order to make the code more readable and reusable we will define the following expressions:
+	// y = signedWeight^2 + 2^(d + 2) * signedWeight + 2^2d
+	// x = 3 * 2^b * (signedWeight^2 - 2^2d)
+	// w = d * (T - 1)
+	//
+	// numReveals > = ((strengthTarget) * T * Y / (3 * 2^b * (signedWeight^2 - 2^2d) + (d * (T - 1) - P) * Y))
+	//        /\
+	//        ||
+	//        \/
+	// numReveals > = ((strengthTarget) * T * y / (x + (w - P) * y))
+	y, x, w := getSubExpressions(signedWeight)
+
```

Comment:
Adding a pseudocode here will help readability and relating the code with the formula in the comments. 
```suggestion
        // numerator = strengthTarget * ln2IntApproximation * y
```